compat/memset_s.cc

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#if !HAVE_MEMSET_S
+
+#include "compat/memset_s.h"
+
+/**
+ * to zero a buffer in a more secure manner meant for a handful of purposes.
+ * e.g. for password matters.
+ * The compiler can optimize away a memset call to gain performance here
+ * making sure it does not occur.
+ *
+ * memset_s is designed for such cases otherwise putting the
+ * address in a volatile pointer avoid gcc's likes doing optimizations.
+ */
+errno_t memset_s(void *dst, rsize_t dsz, int c, rsize_t len)
+{
+    errno_t ret = 0;
+    if (!dst)
+	    return EINVAL;
+    if (dsz > SIZE_MAX)
+	    return E2BIG;
+    if (len > dsz) {
+	    len = dsz;
+	    ret = EOVERFLOW;
+    }
+    void *(*volatile memset_fn)(void *, int, size_t) = &memset;
+    (void)memset_fn(dst, c, len);
+    return ret;
```

Comment:
This implementation does not properly replace a [standard](https://en.cppreference.com/w/c/string/byte/memset) memset_s() AFAICT. The standard one calls a constraint handler function upon detecting violations, right? Whether we should implement that aspect depends on the (currently unknown to me) PR goals/intent, but we should definitely disclose such a serious implementation deficiency.

N.B. [C11](https://en.cppreference.com/w/c/error/abort_handler_s) says that "the default handler is implementation-defined: it may be abort_handler_s, ignore_handler_s, or some other implementation-defined handler".

Reply:
here it does the bare minimum, not a real replacement since it is not meant to be generally widely used, protections are mainly to protect against possible bad usage in general but in squid they re properly done, the main point is the placement of memset address into a volatile address. System implementations implement these part in assembly for example but that might be a bit far stretched here :-).